lib/UserCohort.js

```diff
@@ -0,0 +1,81 @@
+const mongoose = require("./utils/mongoose");
+const Schema = mongoose.Schema;
+const Cohort = require("./Cohort");
+const User = require("./User");
+
+const UserCohort = new Schema(
+  {
+    user: {
+      type: Schema.Types.ObjectId,
+      ref: "User",
+      required: "{PATH} is required!",
+      unique: true
+    },
+    cohort: {
+      type: Schema.Types.ObjectId,
+      ref: "Cohort"
+    }
+  },
+  { timestamps: true }
+);
+UserCohort.plugin(require("./plugins/mongoose-to-id"));
+UserCohort.plugin(require("mongoose-findorcreate"));
+UserCohort.plugin(require("./plugins/mongoose-no-underscore-id"));
+
+UserCohort.statics.findForUser = async function(user) {
+  if (!(user instanceof User)) {
+    throw new Error("user param must be a User instance");
+  }
+  return await this.findOne({ user: user.id }).exec();
+};
+
+UserCohort.statics.setUserCohort = async function(user, cohortName) {
+  if (!(user instanceof User)) {
+    throw new Error("user param must be a User instance");
+  }
+
+  const userCohortModel = this;
+  const cohort = await Cohort.findOrCreateForName(cohortName);
+  let userCohort = await userCohortModel.findForUser(user);
+  if (userCohort && `${cohort}` === `${userCohort.cohort}`) {
+    return userCohort;
+  }
+
+  userCohort = await userCohortModel
+    .findOneAndUpdate(
+      { user: user.id },
+      {
+        user: user.id,
+        cohort: cohort.id
+      },
+      { upsert: true, new: true }
+    )
+    .exec();
+
+  // Check if user is in any goal cohorts
+  const goalCohortModel = mongoose.model("GoalCohort");
+  const goalCohorts = await goalCohortModel.find({
+    members: {
+      $elemMatch: {
+        user: user.id
+      }
+    }
+  });
+
+  // Leave old cohorts
+  const goalModel = mongoose.model("Goal");
+  for (const goalCohort of goalCohorts) {
+    const goal = await goalModel.findById(goalCohort.goal);
```

Comment:
we should do two changes for time performance here:

1. find all the goals in a single query, e.g.

```js
// query may not be exactly right
const goalList = await goalModel.find({_id: { $in: goalCohorts.map(g => g.goal) } });
```

2. issue the `leaveCohort` requests as a batch. I don't think we actually need to await the results before returning, but if we want to await them just to log a potential error could be like this:

```js
const leavePromises = goalList.map(g => goalCohortModel.leaveCohort(user, goal));
Promise.all(leavePromises).catch(err => logging.error(err);
return userCohort;
```